### DISCUSSION TITLE
Prevent scrollbar from causing side shift?

### DISCUSSION THREAD

**Discussion by twMat** [TiddlyWiki/TiddlyWiki5#6350] [General]
No biggie but if you close up tiddlers in the river until they don't reach below the viewport, then the browser *scrollbar* disappears which, with "fluid story, fixed sidebar", makes the sidebar jump to fill up the space. And if some tiddler is opened again, the sidebar jumps back into existance, shifting the position of the sidebar to make room. This is just not visually elegant. The jumping does *not* happen for the second scollbar, i.e for the "inner" scrollbar that appears if the sidebar listing itself is too long. 

Could this shift-from-scrollbar be avoided? Maybe it is a simple CSS matter?

**Comment by linonetwo** [comment]
Yes, in chrome you can use css to hide all scrollbars,while still able to scroll.

But this won't work in firefox and safari.
but it always works in Electron.

```css
&::-webkit-scrollbar {
    width: 0;
  }
```

This will work.

**Comment by kookma** [comment]
@twMat - I believe scrollbars in Tiddlywiki need some attentions!
What you said, in my opinion is a UI design issue  and I really wish to be addressed by core css!

**Comment by Jermolene** [comment]
I am afraid that the issues with multiple scroll bars appearing with the default TiddlyWiki layout are pretty much due to my use of Apple devices, which by default don't show visible scrollbars, relying on the mouse wheel or scrolling gestures instead. Perhaps I would have chosen different tradeoffs for the vanilla theme if I had been using different devices.

I hadn't come across @linonetwo's trick but it would make sense for people with an input device that supports scrolling gestures.

A core tweak that might be worth exploring would be to provide an option to remove or disable the sidebar scrollable, so that the sidebar only scrolls as part of the main page, as in TiddlyWiki Classic.

**Comment by pmario** [comment]
I did create a plugin with a bit of CSS that makes the scrollbars thinner on windows: https://wikilabs.github.io/editions/thin-scrollbars/ but it doesn't prevent the "jumpyness" behaviour

**Comment by kookma** [comment]
What do you think if a width dedicated to story river scrollbar and prevent sidebar jump to fill up the space?

![img_754_%pn](https://user-images.githubusercontent.com/830394/146771651-2e5933eb-6c6e-49b7-baf5-7e79895620fe.gif)

@linonetwo may provide better screenshots

**Comment by linonetwo** [comment]
There is a feature: `overflow: overlay`, see https://developer.mozilla.org/en-US/docs/Web/CSS/overflow

> overlay
> Behaves the same as auto, but with the scrollbars drawn on top of content instead of taking up space. Only supported in WebKit-based (e.g., Safari) and Blink-based (e.g., Chrome or Opera) browsers.

I learned this today from my colleague.

**Reply by twMat** [reply]
@linonetwo wrote

>overlay
Behaves the same as auto, but with the scrollbars drawn on top of content instead of taking up space. 

This! Yes, IMO there is need for the scrollbar (so the initial replies here to *remove* it confuses me... but if that is Apple behaviour I guess it has merit). 

Here's an attempt to use it:

1. go to tiddlywiki.com
2. close all tids except a long one, e.g the bottom one "Community"
3. make a quick edit to it so you can easily find it in tab Recenwt
4. ...and also leave the tab Recent active so the inner scrollbar is also seen
4. Right click the outer scrollbar to open inspector tool and add `overflow: overlay`

It does work but what happens is it overlays the inner scrollbar, making the inner scrollbar inaccessible. (Maybe `overflow: overlay` is similar to `position: absolute` ?) So the position of the inner scrollbar would need to depend on if the outer is there or not, i.e if the document is overflown or not.

**Comment by twMat** [comment]
It turns out this is a recognized problem - and it has a designated css solution; the `scrollbar-gutter` property:

1. [css-tricks](https://css-tricks.com/almanac/properties/s/scrollbar-gutter/)
2. [bram.us](https://www.bram.us/2021/07/23/prevent-unwanted-layout-shifts-caused-by-scrollbars-with-the-scrollbar-gutter-css-property/)

This is intended to replace the `overflow: overlay` property, which is [deprecated](https://developer.mozilla.org/en-US/docs/Web/CSS/overflow). Unfortunately, the browser coverage is pretty poor for both `overlay` ([73%](https://caniuse.com/css-overflow-overlay)) and `scrollbar-gutter` ([64%](https://caniuse.com/?search=scrollbar-gutter)), and the latter is still at the experimental stage.

Perhaps the easiest solution is to just do `body {overflow-y:scroll !important;}` which works and, for an empty story river, gives an empty scrollbar. I tried to make its background transparent (`body::-webkit-scrollbar-track {background: transparent !important;}`) until hovering, but fail for some reason.

**Comment by linonetwo** [comment]
Combine `:hover` with `{ overflow: overlay }` will have a better effect!

https://github.com/tiddly-gittly/tiddlywiki-plugins/blob/1a5841af3441aff1d037de50d34d04bd04dc786e/plugins/linonetwo/template-list/style.css.tid#L25-L32

I'm using this technique in my https://talk.tiddlywiki.org/t/apply-template-to-your-new-tiddler-a-template-list-plugin/2121 

When hovered:

<img width="957" alt="截屏2022-01-19 01 04 09" src="https://user-images.githubusercontent.com/3746270/149984135-cad2b0d3-fabe-4334-8533-81bf15672bec.png">

When not hovered:

<img width="970" alt="截屏2022-01-19 01 04 18" src="https://user-images.githubusercontent.com/3746270/149984183-0b0c8b3f-0ed9-49dc-8012-9e7b6ae42ba1.png">

**Reply by twMat** [reply]
>Combine :hover with { overflow: overlay } will have a better effect!

@linonetwo - that's neat but it makes the "_sidebar_ scroller", i.e for long sidebar lists, disappear behind the overlaying main scroller.
